Atomic.md

```diff
@@ -0,0 +1,253 @@
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   We can simply retry and have this succeed. Nothing was created
+        in .staging
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We can simply retry and have this succeed. The file written in
+        .staging will be overwritten/deleted and recreated
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have an incomplete placeholder file in .staging that will be
+        overwritten by a retry.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have an incomplete placeholder file in .staging that will be
+        overwritten by a retry.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   We have a partially created placeholder written into .staging that
+        will be overwritten on retry
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Directory is marked as in virtualization root but not empty.
+
+        -   An incomplete placeholder is in .staging that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .staging yet
+
+    -   The rename() to copy the completed placeholder from .staging into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+With this approach, each placeholder file that will be created will be moved in one at a time to their final location rather than moving all placeholders that may be present in a directory at once. 
+
+In a case where we crash (or the machine crashes, or the user kills the mount process) while expanding a placeholder directory, we will be able to recover by calling UpdatePlaceholderInformation on files that were enumerated by the previous attempt and creating the rest of the placeholders as expected. 
+
+We need to call UpdatePlaceholderInformation in order to prevent data loss in scenarios where users manage to create files in the partially enumerated directory (if the kext is unmounted, the user can write files into the directory that will look like full files to VFSForGit).
+UpdatePlaceholderInformation currently will add any file to ModifiedPaths (which will make 'git status' begin tracking it, if any changes are present) and will one day be smart enough to only update files that are still placeholders.
```

Comment:
Small correction here:

`UpdatePlaceholderIfNeeded` will (in the future) fail if the file is not currently a placeholder.  

If this happens then VFSForGit's expansion code (that is calling `UpdatePlaceholderIfNeeded`) will need to update ModifiedPaths.

(It's currently the code for updating the projection that handles failures of `UpdatePlaceholderIfNeeded` and adds paths to ModifiedPaths)